### PR TITLE
add support for custom vocabulary

### PR DESCRIPTION
**Pull request by smdesai**
### Why is this change needed?
Custom lexicons exist to let users override pronunciation for domain-specific terminology that general-purpose G2P and built-in dictionaries either do not support of mishandle —this is especially common in finance (tickers, company names, jargon like “EBITDA”), healthcare (drug names, procedures, acronyms), legal (Latinisms, case citations), and tech (product names, abbreviations). The goal is for these overrides to be dependable and to take effect exactly when the user expects resulting in fluid audio generation.

### CODE REVIEW

File: Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift [hunk 1]
```diff
@@ -0,0 +1,287 @@
+import Foundation
+
+/// A custom pronunciation dictionary for TTS synthesis.
+///
+/// Custom lexicons allow domain-specific pronunciation overrides that take precedence
+/// over all built-in dictionaries and grapheme-to-phoneme conversion.
```
Comment (Alex-Wengg) [inline Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift:6]: do you think its worth adding a doc .md file instead

Comment (smdesai) [reply Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift:6]: Yes I believe it is, a doc as in an .md file, right? I'll take a look and provide something useful.

Comment (smdesai) [reply Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift:6]: Added the doc which contains python code I used for generating phonemes used for testing. I figured it'd be useful for anyone creating it and then modify it afterwards to get the desired behavior.

Comment (Alex-Wengg) [reply Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift:6]: i think we can remove these comments in this swift file then .

for the py code would it complex ? dependencies can be a huge issue

File: Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+/// UN=junˈITᵻd nˈAʃənz
+/// ```
+///
+/// ## Example Usage
+/// ```swift
+/// // From file
+/// let lexicon = try TtsCustomLexicon.load(from: fileURL)
+///
+/// // From string
+/// let lexicon = try TtsCustomLexicon.parse("""
```
Comment (Alex-Wengg) [inline Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift]: unit test for TtsCustomLexicon.swift ? 

Parsing valid/invalid files
Three-tier matching behavior (exact, case-insensitive, normalized)
Unicode grapheme cluster tokenization
merged(with:) conflict resolution
Edge cases: empty files, malformed lines, = in phonemes

File: Sources/FluidAudioCLI/Commands/TTSCommand.swift [hunk 3]
```diff
@@ -192,7 +199,22 @@ public struct TTS {
             // Timing buckets
             let tStart = Date()
 
-            let manager = TtSManager()
+            // Load custom lexicon if provided
+            var customLexicon: TtsCustomLexicon? = nil
```
Comment (Alex-Wengg) [inline Sources/FluidAudioCLI/Commands/TTSCommand.swift]: TTSCommand.swift have the same 12-line lexicon loading block duplicated in both run() and runBenchmark():

File: Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift (hunk 2 above)
Comment (smdesai) [reply Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift]: Yes my bad, it's included now.

File: Sources/FluidAudioCLI/Commands/TTSCommand.swift (hunk 3 above)
Comment (smdesai) [reply Sources/FluidAudioCLI/Commands/TTSCommand.swift]: Refactored.

File: Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift (hunk 1 above)
Comment (smdesai) [reply Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift:6]: Removed the comments from the swift code. The Python code is there as a reference. If someone wants to try it out, it's on them to install the dependencies.

Comment (Alex-Wengg) [reply Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift:6]: i am confused don't we already do these in swift ? EspeakG2P + PhonemeMapper. why do we need python misaki exactly .

Comment (smdesai) [reply Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift:6]: Yes they're in Swift. The Python code is just an example which I found useful. I'm happy to remove it from the doc if it causes confusion.

Comment (Alex-Wengg) [reply Sources/FluidAudioTTS/TextToSpeech/Kokoro/TtsCustomLexicon.swift:6]: i will leave it up to you but do clarify it is optional & swift supports it

File: Documentation/CustomPronunciationDictionary.md [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+### Finding Valid Phonemes
+
+The Kokoro vocabulary uses a specific phoneme set. To find valid phonemes:
+
+1. Look at existing entries in the built-in lexicon
+2. Use eSpeak-NG's IPA output as a reference
+3. Test with short phrases to verify pronunciation
+
+### Phoneme Generator
+The Swift code handles the mapping of IPA tokens to Kokoro vocabulary via PhonemeMapper and EspeakG2P and the Python code here is purely an example of generating and experimenting with phonemes. It serves no other purpose. The code is based on [this](
```
Comment (Alex-Wengg) [inline Documentation/CustomPronunciationDictionary.md]: if it serves no other purposes than lets remove it

Comment (smdesai) [reply Documentation/CustomPronunciationDictionary.md]: It's removed.

**Comment by Alex-Wengg** [comment]
CleanShot 2025-12-14 at 11  41 44@2x


@smdesai

**Review by Alex-Wengg — APPROVED** [review]

**Review by Alex-Wengg — COMMENTED** [review]
@smdesai did you checked if xcode supports your latest change ?

**Comment by smdesai** [comment]
@Alex-Wengg Yes it's tested with Xcode and runs on iOS:

https://github.com/user-attachments/assets/2cf09d8d-e7af-4c3d-8432-6c93cfd97f9a
